### PR TITLE
fix mypy typing errors in pytorch_lightning/tuner/lr_finder.py

### PR DESCRIPTION
**Pull request by donlapark**
## What does this PR do?

Fixes mypy typing errors in `pytorch_lightning/tuner/lr_finder.py` in #13445.

* Import of `matplotlib` has been moved to the beginning of the file in order to enable `plt.Figure` type.
* Line 129:  The type of `scheduler` in `LRSchedulerConfig` is expected to be `Union[_LRScheduler, ReduceLROnPlateau]`, but both `_LinearLR` and `_ExponentialLR` inherit from `torch.optim.lr_scheduler._LRScheduler`.
* Line 209: sometimes `mypy` will throw `Return value expected` error when it detects a lone `return` (https://github.com/python/mypy/issues/1003), in which case I added `# type: ignore` to the line.
* Line 231: `mypy` will throw in an error if a function is assigned to a method (https://github.com/python/mypy/issues/2427).
* Line 332: `Union[_LRScheduler, ReduceLROnPlateau]` has no attribute `lr`. 

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/tuner/lr_finder.py [hunk 1]
```diff
@@ -16,8 +16,9 @@
 import os
 import uuid
 from functools import wraps
-from typing import Any, Dict, Optional, Sequence
+from typing import Any, Callable, Dict, List, Optional, Sequence, Union
 
+import matplotlib.pyplot as plt
```
Comment (akihironitta) [inline src/pytorch_lightning/tuner/lr_finder.py:21]: `matplotlib` is an optional dependency, and thus we cannot assume that it's always available, so we need to protect the import statement like this:

https://github.com/Lightning-AI/lightning/blob/cf189bdc9b32326ea02b9af6dd6e1bf04f0e4d2e/src/pytorch_lightning/strategies/deepspeed.py#L56-L58

File: src/pytorch_lightning/tuner/lr_finder.py [hunk 2]
```diff
@@ -95,16 +96,16 @@ def __init__(self, mode: str, lr_min: float, lr_max: float, num_training: int):
         self.lr_max = lr_max
         self.num_training = num_training
 
-        self.results = {}
+        self.results = {}  # type: Dict[str, Any]
```
Comment (akihironitta) [inline src/pytorch_lightning/tuner/lr_finder.py:99]: Shall we use the following syntax for consistency in the codebase? (kind of related to [PEP526](https://peps.python.org/pep-0526/))
```suggestion
        self.results: Dict[str, Any] = {}
```

**Review by akihironitta — COMMENTED** [review]
Thank you for working on this!

File: src/pytorch_lightning/tuner/lr_finder.py [hunk 3]
```diff
@@ -196,7 +196,7 @@ def lr_find(
     """See :meth:`~pytorch_lightning.tuner.tuning.Tuner.lr_find`"""
     if trainer.fast_dev_run:
         rank_zero_warn("Skipping learning rate finder since fast_dev_run is enabled.")
-        return
+        return  # type: ignore
```
Comment (akihironitta) [inline src/pytorch_lightning/tuner/lr_finder.py:199]: > * Line 199: sometimes `mypy` will throw `Return value expected` error when it detects a lone `return` (https://github.com/python/mypy/issues/1003), in which case I added `# type: ignore` to the line.

Thanks for the information! How about returning `None` explicitly following PEP8 instead?

```suggestion
        return None
```

File: src/pytorch_lightning/tuner/lr_finder.py (hunk 1 above)
Comment (donlapark) [reply src/pytorch_lightning/tuner/lr_finder.py:21]: Thanks! I tried to replicate the same code for Matplotlib in 7645add.

Comment (donlapark) [reply src/pytorch_lightning/tuner/lr_finder.py]: Whoops, I forgot to remove the actual import. It should be fixed now in 4d6068cd1300e4a991d36352d8111f175d9568ac.

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Comment by donlapark** [comment]
Sorry for the closing and reopening PR -- most of the checks are stuck with `Expected — Waiting for status to be reported` so I am trying to unstuck it.

**Comment by justusschock** [comment]
> Sorry for the closing and reopening PR -- most of the checks are stuck with Expected — Waiting for status to be reported so I am trying to unstuck it.

That is because as a first-time contributor you need approval from our team to run our CI (which I just approved)

File: src/pytorch_lightning/tuner/lr_finder.py [hunk 4]
```diff
@@ -16,7 +16,7 @@
 import os
 import uuid
 from functools import wraps
-from typing import Any, Dict, Optional, Sequence
+from typing import Any, Callable, cast, Dict, List, Optional, Sequence, Union
```
Comment (justusschock) [inline src/pytorch_lightning/tuner/lr_finder.py:19]:
```suggestion
from typing import Any, Callable, cast, Dict, List, Optional, Sequence, Union, TYPE_CHECKING
```

File: src/pytorch_lightning/tuner/lr_finder.py [hunk 5]
```diff
@@ -38,6 +39,10 @@
 else:
     from tqdm import tqdm
 
+_MATPLOTLIB_AVAILABLE: bool = _RequirementAvailable("matplotlib")  # type: ignore[assignment]
+if _MATPLOTLIB_AVAILABLE:
```
Comment (justusschock) [inline src/pytorch_lightning/tuner/lr_finder.py:43]:
```suggestion
_MATPLOTLIB_AVAILABLE: bool = _RequirementAvailable("matplotlib")  # type: ignore[assignment]
if _MATPLOTLIB_AVAILABLE and TYPE_CHECKING:
```

File: src/pytorch_lightning/tuner/lr_finder.py [hunk 6]
```diff
@@ -131,14 +137,18 @@ def func(trainer):
...
             suggest: if True, will mark suggested lr to use with a red point
 
             show: if True, will show figure
         """
-        import matplotlib.pyplot as plt
+        if not _MATPLOTLIB_AVAILABLE:
+            raise MisconfigurationException(
+                "To use the `plot` method, you must have Matplotlib installed."
+                " Install it by running `pip install -U matplotlib`."
+            )
```
Comment (justusschock) [inline src/pytorch_lightning/tuner/lr_finder.py:151]:
```suggestion
        if not _MATPLOTLIB_AVAILABLE:
            raise MisconfigurationException(
                "To use the `plot` method, you must have Matplotlib installed."
                " Install it by running `pip install -U matplotlib`."
            )
        import matplotlib.pyplot as plt
```

No need to import it globally all the time just for typing

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/tuner/lr_finder.py [hunk 7]
```diff
@@ -38,6 +39,10 @@
 else:
     from tqdm import tqdm
 
+_MATPLOTLIB_AVAILABLE: bool = _RequirementAvailable("matplotlib")  # type: ignore[assignment]
```
[the remark this answers is not in the document]
Comment (otaj) [reply src/pytorch_lightning/tuner/lr_finder.py:42]: There are two things that come to mind - either explicitly cast it to `bool` somehow, or make all the typing for things like `_*_AVAILABLE` of type `_RequirementAvailable`, which can be used in a boolean comparison because of overwritten `__bool__`

[the remark this answers is not in the document]
Comment (donlapark) [reply src/pytorch_lightning/tuner/lr_finder.py:42]: Oh nice. There is also a similar annotation in another file, as mentioned by @akihironitta:

https://github.com/Lightning-AI/lightning/blob/cf189bdc9b32326ea02b9af6dd6e1bf04f0e4d2e/src/pytorch_lightning/strategies/deepspeed.py#L56-L58

Should we remove the annotation as well?

[the remark this answers is not in the document]
Comment (carmocca) [reply src/pytorch_lightning/tuner/lr_finder.py:42]: Yes, please

[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/tuner/lr_finder.py:42]: Thanks!
